lib/AstOfLlbc.ml

```diff
@@ -385,29 +359,198 @@ let lookup_field env typ_id field_id =
+
+and metadata_typ_of_ty (env : env) (ty : Charon.Types.ty) : K.typ option =
+  match ty with
+  | C.TAdt ty_decl_ref -> begin
+      match ty_decl_ref.id with
+      | C.TAdtId decl_id -> begin
+          let decl = env.get_nth_type decl_id in
+          match decl.ptr_metadata with
+          | C.NoMetadata -> None
+          | C.Length -> Some Krml.Helpers.usize
+          | C.VTable ty_ref ->
+              let ty_ref = { ty_ref with generics = ty_decl_ref.generics } in
+              Some (K.TBuf (typ_of_ty env (C.TAdt ty_ref), false))
+          | C.InheritFrom (C.TVar (C.Free var)) ->
+              let ty = List.nth ty_decl_ref.generics.types (C.TypeVarId.to_int var) in
+              metadata_typ_of_ty env ty
+          | C.InheritFrom _ ->
+              failwith
+                "Eurydice does not handle PtrMetadata inheritance, please consider using \
+                 monomorphized LLBC"
+        end
+      | C.TTuple -> begin
+          match List.rev @@ ty_decl_ref.generics.types with
+          (* Empty metadata for empty tuple *)
+          | [] -> None
+          (* For tuple, the type of metadata is the last element *)
+          | last :: _ -> metadata_typ_of_ty env last
+        end
+      | C.TBuiltin C.TBox | C.TBuiltin C.TArray -> None
+      | C.TBuiltin C.TSlice | C.TBuiltin C.TStr -> Some Krml.Helpers.usize
+    end
+  | C.TVar _ ->
+      let _var_is_sized (clause : C.trait_param) =
+        let trait = clause.trait.binder_value in
+        match trait.generics.types with
+        | [] -> failwith "Unexpected empty `Self` from trait clause."
+        (* The variable here in C.TVar should be free variable, hence no need to adjust the DB id. *)
+        | self :: _ ->
+            self = ty
+            &&
+            let decl = env.get_nth_trait_decl trait.id in
+            (* It has the Sized marker, which is of lang_item "sized" *)
+            decl.item_meta.lang_item = Some "sized"
+      in
+      let var_is_sized =
+        true
+        (*List.exists var_is_sized env.generic_params.trait_clauses*)
+      in
+      if var_is_sized then
+        None
+      else
+        failwith
+          "Eurydice does not handle taking metadata from non-Sized type vars, please consider \
+           using monomorphized LLBC."
+  | C.TTraitType (_, _) ->
+      failwith
+        "Eurydice does not handle taking metadata from assoc types, please consider using \
+         monomorphized LLBC."
+  (* The metadata of a &dyn Trait object is a pointer to its vtable *)
+  | C.TDynTrait pred -> Some (K.TBuf (vtable_typ_of_dyn_pred env pred, false))
+  | C.TLiteral _ | C.TRef (_, _, _) | C.TRawPtr (_, _) | C.TFnPtr _ | C.TFnDef _ -> None
+  (* The metadata must not have ptr metadata as they must be Sized. *)
+  | C.TPtrMetadata _ -> None
+  | C.TNever | C.TError _ -> failwith "Error types to fetch metadata"
+
+(* Get the pointer type of the original Charon type. *)
+and ptr_typ_of_ty (env : env) (ty : Charon.Types.ty) : K.typ =
```

Comment:
suggest different comment: "Translate Charon type &T as a krml type -- this handles special cases where address-taking in Rust creates a DST, which we represent as instances of the `dst_ref` type. There are many such cases: slices, &str, etc."